### PR TITLE
workflows: Run PR workflows on post-merge repo state

### PR DESCRIPTION
**Pull request by bduffany**
The goal is to run workflows on the state of the repo after merging the main branch, to detect potential incompatibilities early, and to pull in compatibility fixes when re-running workflows (without needing to merge the main branch). This behavior is the same as what GitHub actions does with their `checkout` action.

* This would be easy if we could just check out GitHub's "merge refs" that it creates: `refs/pull/{pr_number}/merge`. However it seems like we should not use these -- GitHub folks described these refs as an implementation detail for GitHub that should not be relied on. BitBucket says the same about their merge refs, and worse -- they don't provide any guarantees about when these merge refs will be created.
* Since we can't rely on merge refs, we can just do the merge ourselves by checking out the pushed branch and merging in the target branch. Unfortunately, `go-git` does not support merging. So, this PR migrates everything to using raw `git` commands instead of using `go-git`. (This might actually be a good thing because the `git` binary seems to do fetches much more quickly than `go-git`, and in my experience so far, `go-git` has been kind of a pain to work with since it models things quite a bit differently than the git CLI does).
  * This makes the CI runner non-hermetic, but this is OK because the CI runner image has `git` installed. In the future, when we support custom images, we'll document that `git` needs to be installed.
  * Note, this change will _not_ cause our tests to break, at least in any environments where they are currently passing -- `ci_runner_test` was already non-hermetic because it turns out that `go-git` falls back to invoking the `git` binary for `file://` repo URLs.
  * I tried building a statically linked `git` binary from source but it was a hassle. It depends on a few libs which don't have a static lib distribution available on `apt`. Might revisit this in the future if we run into issues due to git.
* In order to do the merge ourselves, the CI runner needs a little more information (in particular, the "pushed" and "target" repo URLs). This resulted in a lot of "plumbing" of various proto fields.

Since we're using `git` commands for everything, we can now show the actual git commands that are being run , to make workflows more transparent (and make debugging easier). Screenshot:

image

---

**Version bump**: Patch

**Related issues**: Fixes https://github.com/buildbuddy-io/buildbuddy/issues/617, Fixes https://github.com/buildbuddy-io/buildbuddy-internal/issues/693, Fixes https://github.com/buildbuddy-io/buildbuddy-internal/issues/695

### CODE REVIEW

File: proto/build_event_stream.proto [hunk 1]
```diff
@@ -827,6 +824,21 @@ message WorkflowConfigured {
...
+  string pushed_branch = 7;
+
+  string commit_sha = 8;
+
+  // The repo into which the pushed branch is being merged (for PRs),
+  // or the pushed_repo_url (for push events).
+  // Ex: "https://github.com/acme-inc/acme"
+  string target_repo_url = 9;
+
+  string target_branch = 10;
```
Comment (tylerwilliams) [inline proto/build_event_stream.proto:841]: nit: comment these or remove the newlines  between fields if the comment applies to multiple

File: proto/workflow.proto [hunk 2]
```diff
@@ -131,9 +131,24 @@ message ExecuteWorkflowRequest {
...
+
   // Branch corresponding to the SHA to be checked out.
   // Ex. "cool-feature"
-  string branch = 5;
+  string pushed_branch = 7;
+
+  // URL of the target repo.
+  // Ex. "https://github.com/acme-inc/acme"
+  string target_repo_url = 8;
+
```
Comment (tylerwilliams) [inline proto/workflow.proto:147]: nit: comment on pushed_repo* and target_repo* what these are:
 pushed repo is the repository this commit was pushed to
 target repo is the repository this commit will be merged into.

or something similar?

File: enterprise/server/cmd/ci_runner/main.go [hunk 3]
```diff
@@ -99,15 +100,28 @@ var (
...
-	// initLogs contain informational logs from the setup phase (cloning the
-	// git repo and deciding which actions to run) which are reported as part of
-	// the first action's logs.
-	initLog bytes.Buffer
 )
 
+type workspace struct {
+	// Whether the pushed branch could not be cleanly merged into the targt
+	// branch.
+	mergeConflict bool
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:108]: ubernit: flip to mergesCleanly ("positive" variables are easier to read and grok)

Then you can have if !mergesCleanly (if not merges cleanly)?

File: enterprise/server/cmd/ci_runner/main.go [hunk 4]
```diff
@@ -499,27 +502,34 @@ func (ar *actionRunner) Run(ctx context.Context, startTime time.Time) error {
...
+				BazelCommand: bazelCmd,
+			})
+			eventID := &bespb.BuildEventId{
+				Id: &bespb.BuildEventId_WorkflowCommandCompleted{WorkflowCommandCompleted: &bespb.BuildEventId_WorkflowCommandCompletedId{
+					InvocationId: iid,
+				}},
+			}
+			configuredEvent.Children = append(configuredEvent.Children, eventID)
 		}
-		configuredEvent.Children = append(configuredEvent.Children, eventID)
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:522]: this is a bit confusing: the comment says "if merge conflict", the code says "if not merge conflict", and for someone not familiar with this code, it's tough to understand the relation between the bazel commands the comment references and what happens in the loop (sending invocation events?)

maybe you could say something like:

// If this commit merges cleanly, the runner will execute several bazel commands below to run tests. If the commit does not merge cleanly, the runner will exit early without running those commands and does not need to create invocation streams for them.

or something similar (you know what's going on here way better than I)

File: enterprise/server/cmd/ci_runner/main.go [hunk 5]
```diff
@@ -543,19 +553,21 @@ func (ar *actionRunner) Run(ctx context.Context, startTime time.Time) error {
...
-				// status command that they've configured for bazel.
 			},
 		}},
 	}
 	if err := bep.Publish(workspaceStatusEvent); err != nil {
 		return nil
 	}
 
+	if ws.mergeConflict {
+		return fmt.Errorf("Merge conflict (merging branch %q into %q)", *pushedBranch, *targetBranch)
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:568]: looking at this code i see status errors returned -- is there a reason to return this as a fmt.Errorf?

File: enterprise/server/cmd/ci_runner/main.go [hunk 6]
```diff
@@ -753,79 +766,117 @@ func setupGitRepo(ctx context.Context) error {
...
+	// TODO: Fetch from remotes in parallel
+	baseBranches := []string{*targetBranch}
+	forkBranches := []string{}
+	if *pushedRepoURL == *targetRepoURL {
+		if *pushedBranch != *targetBranch {
+			baseBranches = append(baseBranches, *pushedBranch)
+		}
+	} else {
+		forkBranches = append(forkBranches, *pushedBranch)
+	}
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:795]: A comment here would help clarify what this is doing / why

File: enterprise/server/cmd/ci_runner/main.go [hunk 7]
```diff
@@ -753,79 +766,117 @@ func setupGitRepo(ctx context.Context) error {
...
-		return status.WrapErrorf(err, "checkout %s", *commitSHA)
+	fetchArgs := append([]string{"fetch", "--force", remoteName}, branches...)
+	if err := git(ctx, &ws.log, fetchArgs...); err != nil {
+		return status.WrapError(err, "git fetch")
 	}
 	return nil
 }
 
+func isRemoteAlreadyExists(err error) bool {
+	return getExitCode(err) == 128
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:869]: nit: link to where these codes are defined

File: enterprise/server/cmd/ci_runner/main.go [hunk 8]
```diff
@@ -877,6 +928,14 @@ func getDefaultConfig() *config.BuildBuddyConfig {
 	}
 }
 
+func gitRemoteName(repoURL string) string {
+	if repoURL == *targetRepoURL {
+		return defaultGitRemoteName
+	}
+	shaHex := fmt.Sprintf("%x", sha256.Sum256([]byte(repoURL)))
+	return fmt.Sprintf("fork-%s", shaHex[:8])
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:936]: why 8?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/cmd/ci_runner/main.go (hunk 7 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:869]: Looks like these codes aren't special -- git seems to use exit codes 1 and 128 somewhat arbitrarily. Switched to checking the git command output instead, although both approaches seem equally brittle.

File: enterprise/server/cmd/ci_runner/main.go (hunk 4 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:522]: Done

**Comment by tylerwilliams** [comment]
(reverse-ping -- happy to look at this again but seems like right now it has conflicts and failing tests)

**Comment by bduffany** [comment]
Tests are passing, but I'm seeing an issue where bazel cannot find an external repository after switching branches. Might be an issue with bazel itself, because `bazel clean` solves the problem. Might be able to work around by using a different strategy to check out the branch.

**Comment by bduffany** [comment]
Looks like the above problem I mentioned above is specific to the example repo I was using (and happens even if I just run `bazel test //...` twice in a row). So I think this is good to go

File: enterprise/server/cmd/ci_runner/main.go [hunk 9]
```diff
@@ -129,33 +147,22 @@ func main() {
...
+	if err := ws.setup(ctx); err != nil {
 		fatal(status.WrapError(err, "failed to set up git repo"))
 	}
-	if err := os.Chdir(repoDirName); err != nil {
-		fatal(status.WrapErrorf(err, "cd %q", repoDirName))
-	}
-	runDebugCommand(&initLog, `pwd`)
-	runDebugCommand(&initLog, `ls -la`)
+	runDebugCommand(ws.log, `pwd`)
+	runDebugCommand(ws.log, `ls -la`)
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:154]: can we remove the runDebugCommand stuff now?

File: enterprise/server/cmd/ci_runner/main.go [hunk 10]
```diff
@@ -210,14 +217,19 @@ func RunAllActions(ctx context.Context, cfg *config.BuildBuddyConfig, im *initMe
...
-		if !im.reported {
-			im.reported = true
-			ar.log.Printf("Fetched Git repository in %s\n", startTime.Sub(im.start))
-			startTime = im.start
+		if !ws.reportedInitMetrics {
+			ws.reportedInitMetrics = true
+			// Print the setup logs to the first action's logs.
+			b, _ := io.ReadAll(&ws.logBuf)
+			ws.logBuf.Reset()
+			ar.log.Printf("[BUILDBUDDY] Setup completed in %s", startTime.Sub(ws.startTime))
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:225]: can we remove the "[BUILDBUDDY] " prefix here and below?

File: enterprise/server/cmd/ci_runner/main.go [hunk 11]
```diff
@@ -758,79 +784,152 @@ func setupGitRepo(ctx context.Context) error {
...
+	}
+	return nil
+}
+
+func writeCommandSummary(out io.Writer, format string, args ...interface{}) {
+	io.WriteString(out, ansiGray)
+	io.WriteString(out, "[BUILDBUDDY] ")
+	io.WriteString(out, fmt.Sprintf(format, args...))
+	io.WriteString(out, ansiReset)
+	io.WriteString(out, "\n")
```
Comment (tylerwilliams) [inline enterprise/server/cmd/ci_runner/main.go:930]: you have aurora.Sprintf(aurora.Red("\nAction failed: %s") below -- why not use that here?

**Review by tylerwilliams — CHANGES_REQUESTED** [review]

File: enterprise/server/cmd/ci_runner/main.go (hunk 9 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:154]: Done

File: enterprise/server/cmd/ci_runner/main.go (hunk 10 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:225]: Done

File: enterprise/server/cmd/ci_runner/main.go (hunk 11 above)
Comment (bduffany) [reply enterprise/server/cmd/ci_runner/main.go:930]: The ANSI codes that Aurora uses for gray aren't handled properly by our UI.

**Review by tylerwilliams — APPROVED** [review]
LGTM

**Pull request merged**